driver/src/main/com/mongodb/client/gridfs/GridFSBucket.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.gridfs;
+
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
+import com.mongodb.annotations.ThreadSafe;
+import com.mongodb.client.gridfs.model.GridFSDownloadByNameOptions;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Represents a GridFS Bucket
+ *
+ * @since 3.1
+ */
+@ThreadSafe
+public interface GridFSBucket {
+
```

Comment:
I commented earlier that you shouldn't be able to change the bucketName, as that's really a fundamental change to the bucket itself.  It would be like:

     mongoCollection.withName(String newName)

and we don't allow that.